lib/Runtime/Library/JavascriptSet.cpp

```diff
@@ -6,379 +6,665 @@
+    // TODO: we can use a more efficient Iterator, since we know there will be no side effects
+    while (iter.Next())
+    {
+        varSet->Add(iter.Current(), iter.CurrentNode());
     }
+    return varSet;
+}
 
-    Var JavascriptSet::EntryHas(RecyclableObject* function, CallInfo callInfo, ...)
-    {
-        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+void
+JavascriptSet::PromoteToSimpleVarSet()
+{
+    AssertOrFailFast(this->kind == SetKind::IntSet);
+    SimpleVarDataSet* newSet = this->CreateVarSetFromList<SimpleVarDataSet>(this->u.intSet->Count() + 1);
 
-        ARGUMENTS(args, callInfo);
-        ScriptContext* scriptContext = function->GetScriptContext();
+    this->kind = SetKind::SimpleVarSet;
+    this->u.simpleVarSet = newSet;
+}
 
-        if (!JavascriptSet::Is(args[0]))
-        {
-            JavascriptError::ThrowTypeErrorVar(scriptContext, JSERR_NeedObjectOfType, _u("Set.prototype.has"), _u("Set"));
-        }
+void
+JavascriptSet::PromoteToComplexVarSet()
+{
+    uint setSize = 0;
+    if (this->kind == SetKind::IntSet)
+    {
+        setSize = this->u.intSet->Count();
+    }
+    else
+    {
+        AssertOrFailFast(this->kind == SetKind::SimpleVarSet);
+        setSize = this->u.simpleVarSet->Count();
+    }
+    ComplexVarDataSet* newSet = this->CreateVarSetFromList<ComplexVarDataSet>(setSize + 1);
 
-        JavascriptSet* set = JavascriptSet::FromVar(args[0]);
+    this->kind = SetKind::ComplexVarSet;
+    this->u.complexVarSet = newSet;
+}
 
+void
+JavascriptSet::AddToEmptySet(Var value)
+{
+    // We cannot store an int with value -1 inside of a bit vector
+    Var taggedInt = JavascriptConversion::TryCanonicalizeAsTaggedInt<false>(value);
+    if (taggedInt)
+    {
+        int32 intVal = TaggedInt::ToInt32(taggedInt);
 
-        Var value = (args.Info.Count > 1) ? args[1] : scriptContext->GetLibrary()->GetUndefined();
+        BVSparse<Recycler>* newIntSet = RecyclerNew(this->GetRecycler(), BVSparse<Recycler>, this->GetRecycler());
+        newIntSet->Set(intVal);
 
-        bool hasValue = set->Has(value);
+        this->list.Append(taggedInt, this->GetRecycler());
 
-        return scriptContext->GetLibrary()->CreateBoolean(hasValue);
+        this->u.intSet = newIntSet;
+        this->kind = SetKind::IntSet;
+        return;
     }
 
-    Var JavascriptSet::EntrySizeGetter(RecyclableObject* function, CallInfo callInfo, ...)
+    Var simpleVar = JavascriptConversion::TryCanonicalizeAsSimpleVar(value);
+    if (simpleVar)
     {
-        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+        SimpleVarDataSet* newSimpleSet = RecyclerNew(this->GetRecycler(), SimpleVarDataSet, this->GetRecycler());
+        SetDataNode* node = this->list.Append(simpleVar, this->GetRecycler());
 
-        ARGUMENTS(args, callInfo);
-        ScriptContext* scriptContext = function->GetScriptContext();
+        newSimpleSet->Add(simpleVar, node);
```

Comment:
I like the idea of adding to newIntSet fist and then to this->list as done in AddToEmptySet for case if Append throws then there is no inconsistency.

Reply:
i thought the same. unfortunately i can't really do it here, because i need the node  before adding to the set. i thought i might be able to add a dummy node and then update it, but seemed like might just be better to rely on our OOM failfast to maintain consistency

---
In reply to: [174326068](https://github.com/Microsoft/ChakraCore/pull/4816#discussion_r174326068) [](ancestors = 174326068)